# EDA

In [4]:
import pandas as pd
pd.set_option('display.max_columns', None)

Рассмотрим датасет, полученный скрейпингом с сайта Литрес и обогращенный при помощи api гугл книг

In [18]:
df1 = pd.read_csv("google_x_litres_final.csv")
df1.head()

,Unnamed: 0.1,id_book,categories,publisher,description,print_type,page_count,language,country,saleability,is_ebook,retail_price,currency_code,viewability,public_domain,epub_avaliable,pdf_avaliable,Unnamed: 0,title,author,isbn,pages,publication_year,rating,reviews_count,price,currency,age_restriction,genres,publisher.1,formats,url,price_numeric,rating_numeric,pages_numeric,reviews_numeric
0,0,1Y11EQAAQBAJ,['Fiction'],"ЛитРес, SelfPub",Он был моей единственной любовью. Плохой парен...,NaN,582.0,ru,NL,FOR_SALE,True,2.70,EUR,PARTIAL,False,True,True,0,По осколкам твоего сердца,Анна Джейн,NaN,500.0,2025.0,4.9,3530.0,229.0,RUB,16+,"Литрес Авторы, От ненависти до любви, Первая л...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/anna-dzheyn/po-osko...,229.0,4.9,500.0,3530.0
1,1,Cb-UEAAAQBAJ,['Young Adult Fiction'],"ЛитРес, SelfPub",Я вижу его каждый день в окне дома напротив. О...,NaN,611.0,ru,NL,FOR_SALE,True,2.22,EUR,PARTIAL,False,True,True,1,Твое сердце будет разбито,Анна Джейн,NaN,510.0,2022.0,4.8,7228.0,199.0,RUB,16+,"Литрес Авторы, Любовь и ненависть, Молодежные ...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/anna-dzheyn/tvoe-se...,199.0,4.8,510.0,7228.0
2,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,Фейерверк на ладони,Ольга Назарова,NaN,430.0,2026.0,4.9,139.0,189.0,RUB,16+,"Антистресс, Ироничная проза, Книги для души, К...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/olga-stanislavovna-...,189.0,4.9,430.0,139.0
3,3,p4_kzgEACAAJ,NaN,Ast,NaN,NaN,576.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,NO_PAGES,False,False,False,3,Проект «Аве Мария»,Энди Вейер,NaN,21.0,2021.0,4.8,2117.0,529.0,RUB,16+,"Авантюрные приключения, Близкое будущее, Заруб...",Правообладатель: Аудио-ЛАУ,"m4b,, mp3,, zip",https://www.litres.ru/audiobook/endi-veyer/pro...,529.0,4.8,21.0,2117.0
4,4,1Y11EQAAQBAJ,['Fiction'],"ЛитРес, SelfPub",Он был моей единственной любовью. Плохой парен...,NaN,582.0,ru,NL,FOR_SALE,True,2.70,EUR,PARTIAL,False,True,True,4,По осколкам твоего сердца,Анна Джейн,NaN,14.0,2022.0,4.8,921.0,749.0,RUB,16+,"Young adult, Книги о подростках, Любовь и нена...",Правообладатель: Издательство CLEVER,"m4b,, mp3,, zip",https://www.litres.ru/audiobook/anna-dzheyn/po...,749.0,4.8,14.0,921.0


In [ ]:
df1.dtypes

Удалим заведомо ненужные для анализа столбцы

In [20]:
df1 = df1.drop(columns=["Unnamed: 0.1", "Unnamed: 0", "id_book", "price_numeric", "rating_numeric", "pages_numeric", "reviews_numeric"]) 

In [21]:
df1.shape

(10070, 29)

In [28]:
df1[0:1]

,categories,publisher,description,print_type,page_count,language,country,saleability,is_ebook,retail_price,currency_code,viewability,public_domain,epub_avaliable,pdf_avaliable,title,author,isbn,pages,publication_year,rating,reviews_count,price,currency,age_restriction,genres,publisher.1,formats,url
0,['Fiction'],"ЛитРес, SelfPub",Он был моей единственной любовью. Плохой парен...,NaN,582.0,ru,NL,FOR_SALE,True,2.7,EUR,PARTIAL,False,True,True,По осколкам твоего сердца,Анна Джейн,NaN,500.0,2025.0,4.9,3530.0,229.0,RUB,16+,"Литрес Авторы, От ненависти до любви, Первая л...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/anna-dzheyn/po-osko...


Переставим столбцы в более логичном для анализа порядке и переименуем некоторые столбцы для избежания разночтений

In [32]:
df1 = df1[["title", "author", "isbn", "rating", "reviews_count", "price", "currency", "publication_year", "pages", "age_restriction",
          "genres", "publisher.1", "formats", "url", "categories", "publisher", "description", "print_type", "page_count", "language", "country",
          "saleability", "is_ebook", "retail_price", "currency_code", "viewability", "public_domain", "epub_avaliable", "pdf_avaliable"]]

Рассмотрим полные дубликаты

In [43]:
n_duplicates = df1.duplicated(keep=False)
df1[n_duplicates].head(30)

,title,author,isbn,rating,reviews_count,price,currency,publication_year,pages,age_restriction,genres,publisher.1,formats,url,categories,publisher,description,print_type,page_count,language,country,saleability,is_ebook,retail_price,currency_code,viewability,public_domain,epub_avaliable,pdf_avaliable
647,Влюбленная ведьма,Анна Джейн,NaN,4.9,3536.0,179.0,RUB,2022.0,500.0,16+,"Литрес Авторы, Любовь и ненависть, Новый год, ...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/anna-dzheyn/vlublen...,['College students'],NaN,NaN,NaN,0.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,NO_PAGES,False,False,False
648,Влюбленная ведьма,Анна Джейн,NaN,4.9,3536.0,179.0,RUB,2022.0,500.0,16+,"Литрес Авторы, Любовь и ненависть, Новый год, ...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/anna-dzheyn/vlublen...,['College students'],NaN,NaN,NaN,0.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,NO_PAGES,False,False,False
719,Звездная Кровь – 2. Фригольд,Роман Прокофьев,NaN,4.9,2594.0,159.0,RUB,2023.0,290.0,18+,"LitRPG, Боевая фантастика, Героическое фэнтези...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/roman-urevich-proko...,['Fiction'],ЛитРес,"Оказывается, это не Земля.И не другая планета....",NaN,305.0,ru,NL,FOR_SALE,True,2.16,EUR,PARTIAL,False,True,True
720,Звездная Кровь – 2. Фригольд,Роман Прокофьев,NaN,4.9,2594.0,159.0,RUB,2023.0,290.0,18+,"LitRPG, Боевая фантастика, Героическое фэнтези...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/roman-urevich-proko...,['Fiction'],ЛитРес,"Оказывается, это не Земля.И не другая планета....",NaN,305.0,ru,NL,FOR_SALE,True,2.16,EUR,PARTIAL,False,True,True
1710,Сквотеры,"Влад Лей,",NaN,4.9,129.0,399.0,RUB,2025.0,8.0,16+,"Боевая фантастика, Боевики, Зомби, Книги о при...","Правообладатели: ЛитРес: чтец, Автор","m4b,, mp3,, zip",https://www.litres.ru//audiobook/aleksandr-gro...,['Fiction'],"ЛитРес, SelfPub","Небольшой поселок, в котором осели пережившие ...",NaN,308.0,ru,NL,FOR_SALE,True,2.70,EUR,PARTIAL,False,True,True
1711,Черный север,Татьяна Солодкова,NaN,5.0,192.0,229.0,RUB,2025.0,300.0,16+,"Героическое фэнтези, Детективное фэнтези, Зимн...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru//book/tatyana-solodkova/...,['Fiction'],"ЛитРес, SelfPub","Прошло три года с тех пор, как в Реонерии была...",NaN,382.0,ru,NL,FOR_SALE,True,2.81,EUR,PARTIAL,False,True,True
1714,Джон Стрелеки. Кафе на краю земли. Комплект кн...,Джон П. Стрелеки,978-5-04-197463-3,4.8,45.0,NaN,RUB,NaN,404.0,16+,"Вдохновляющие книги, Жизненные ценности, Заруб...",Правообладатель: Эксмо,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru//book/john-strelecky/dzh...,['Psychology'],ЛитРес,Комплект из четырех культовых книг Джона П. Ст...,NaN,517.0,ru,NL,FOR_SALE,True,14.30,EUR,PARTIAL,False,True,True
1739,ГКР-10: Исход Забытых Душ,Дем Михайлов,NaN,4.8,42.0,176.0,RUB,2025.0,60.0,16+,"LitRPG, Литрес Авторы, Черновик, Юмористическа...",Правообладатель: Автор,текст,https://www.litres.ru//book/dem-mihaylov/gkr-1...,['Fiction'],ЛитРес,Прошли столетья с тех пор как отгремели чудови...,NaN,291.0,ru,NL,FOR_SALE,True,1.95,EUR,PARTIAL,False,True,True
1752,Ложится мгла на старые ступени,Александр Чудаков,978-5-389-22585-5,4.7,200.0,549.0,RUB,2000.0,721.0,18+,"Основано на реальных событиях, Портрет эпохи, ...",Правообладатель: Азбука,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru//book/aleksandr-chudakov...,['Fiction'],Litres,"Александр Чудаков (1938–2005), выдающийся росс...",NaN,745.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,PARTIAL,False,True,True
1753,Ложится мгла на старые ступени,Александр Чудаков,978-5-389-22585-5,4.7,200.0,549.0,RUB,2000.0,721.0,18+,"Основано на реальных событиях, Портрет эпохи, ...",Правообладатель: Азбука,"epub,, fb2,, fb3,, ios.epub,, mobi,, pd

Можно заметить, что многие полные дубликаты идут с последовательными номерами (как например 647 и 647 или 1752 и 1753). Это было замечено на этапе парсинга, когда при загрузке некоторых страниц одна и та же книга считалась первой книгой страницы i и последней книгой страницы i-1. Вероятно, это происходило из-за попытки Литреса динамтчески подгружать следующую страницу.

Другие дубликаты возникли из-за неустойчивости рекомендательной выдачи. Так как датасет парсился по частям, в разное время и разными людьми, то одна и та же книга возникала в выдаче несколько раз и осталась при объединении - такие дубликаты можно спокойно удалить

In [48]:
df1 = df1.drop_duplicates()
df1.shape

(8910, 29)

Рассмотрим другой "тип" дубликатов. Дублирование одного и того же издания, но с разницами в каких-то полях. Тут посмотрим по url.

In [53]:
n_duplicates2 = df1.duplicated(subset=["url"], keep=False)
df1[n_duplicates2]

,title,author,isbn,rating,reviews_count,price,currency,publication_year,pages,age_restriction,genres,publisher.1,formats,url,categories,publisher,description,print_type,page_count,language,country,saleability,is_ebook,retail_price,currency_code,viewability,public_domain,epub_avaliable,pdf_avaliable
1792,С жизнью наедине,Кристин Ханна,978-5-86471-792-9,4.8,2822.0,389.0,RUB,2018.0,461.0,16+,"Аляска, Бестселлеры «New York Times», История ...",Правообладатель: Фантом Пресс,"epub,, fb2,, fb3,, html,, ios.epub,, mobi,, pd...",https://www.litres.ru//book/kristin-hanna/s-zh...,['Fiction'],Phantom Press,"1974-й. Лени тринадцать лет, ее отец недавно в...",NaN,511.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,PARTIAL,False,True,True
1909,С жизнью наедине,Кристин Ханна,978-5-86471-792-9,4.8,2823.0,389.0,RUB,2018.0,461.0,16+,"Аляска, Бестселлеры «New York Times», История ...",Правообладатель: Фантом Пресс,"epub,, fb2,, fb3,, html,, ios.epub,, mobi,, pd...",https://www.litres.ru//book/kristin-hanna/s-zh...,['Fiction'],Phantom Press,"1974-й. Лени тринадцать лет, ее отец недавно в...",NaN,511.0,ru,NL,NOT_FOR_SALE,False,NaN,NaN,PARTIAL,False,True,True
2602,Мара и Морок. Трилогия,Лия Арден,978-5-04-208339-6,4.8,289.0,729.0,RUB,2024.0,919.0,16+,Young adultАвторский мирГероическое фэнтезиДар...,Правообладатель: Эксмо,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/liya-arden/mara-i-m...,['Young Adult Fiction'],ЛитРес,Цикл «Смерть и Тень» Лии Арден теперь в формат...,NaN,1082.0,ru,NL,FOR_SALE,True,7.00,EUR,PARTIAL,False,True,True
2603,Моя гениальная подруга,Элена Ферранте,9785906837479,4.6,3412.0,329.0,RUB,2011.0,330.0,16+,ВзрослениеЖенская дружбаЖенская прозаЖенские с...,Правообладатель: Издательство «Синдбад»,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/elena-ferrante-1041...,['Fiction'],"ЛитРес, Синдбад",Первый из четырех романов уже ставшего культов...,NaN,354.0,ru,NL,FOR_SALE,True,5.21,EUR,PARTIAL,False,True,True
2604,Призраки Марта,Татьяна Солодкова,978-5-9922-3327-8,4.9,421.0,209.9,RUB,2021.0,550.0,16+,Героическое фэнтезиИнтригиКниги про волшебнико...,Правообладатель: АЛЬФА-КНИГА,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/tatyana-solodkova/p...,['Fiction'],"ЛитРес, Альфа-книга",Времена мирного сосуществования черных и белых...,NaN,607.0,ru,NL,FOR_SALE,True,2.70,EUR,PARTIAL,False,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9832,"Начни заканчивать! Иди до конца, действуй и по...",Питер Холлинс,978-5-17-147914-5,4.7,36.0,589.0,RUB,2018.0,100.0,16+,"Как достичь успеха, Как достичь цели, Личная э...",Правообладатель: Издательство АСТ,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/piter-hollins/nachn...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9865,Костры на алтарях,Вадим Панов,978-5-699-25283-1,4.8,454.0,329.0,RUB,2007.0,440.0,16+,"Борьба за власть, Далёкое будущее, Древние арт...","Правообладатели: Эксмо, Автор","epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/vadim-panov/kostry-...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9886,Кривая логика,Александра Шервинская,NaN,4.9,301.0,164.0,RUB,2025.0,200.0,16+,"Городское фэнтези, Детективное фэнтези, Дружба...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/aleksandra-shervins...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9930,Диагностика кармы. Чистая карма. Часть 2,Сергей Лазарев,NaN,4.9,196.0,690.0,RUB,2024.0,230.0,16+,"Внутренний мир человека, Диагностика, Исцеляющ...",Правообладатель: Автор,"epub,, fb2,, fb3,, ios.epub,, mobi,, pdf,, txt...",https://www.litres.ru/book/sergey-nikolaevich-...,"['Body, Mind & Spirit']","ЛитРес, SelfPub",В этот раз популярный философ и автор книг о к...,NaN,245.0,ru,NL,FOR_SALE,True,7.67,EUR,PARTIAL,False,True,True


Иногда отличается кол-во оценок (rev)